main/GUB2.c

```diff
@@ -0,0 +1,309 @@
+    if (ret == ESP_OK) {
+        int r = GUBMountSDCard(SD_CARD_BASE_PATH);
+        ESP_LOGI("SD_CARD", "SD card initialized with status %d", r);
+        if(!r){
+            sdmmc_card_print_info(stdout, gubState.SDcard);
+        }
+    } else {
+        ESP_LOGE(TAG, "Failed to initialize bus.");
+    }
+
+    listDir(SD_CARD_BASE_PATH);
+
+    //start up logger
+    canLoggerInit();
+
+    GUBInitLED();
+
+    ESP_LOGI(TAG, "GUB Setup, starting main loop");
+    xTaskCreate( GUBloop, "GUB", GUB_STACK_SIZE, NULL, 2, &gubState.mainTaskHandler);
+}
+
+/**
+ * Main loop for GUB task
+*/
+void GUBloop(void *pvParam){
+
+    while (true)
+    {   
+        GUBHeartbeatUpdate();
+        checkSDCardStatus();
+        if(gubState.sdCardState == SD_NOT_MOUNTED && esp_timer_get_time() - gubState.lastMountAttempt > 1000000){ // check every second
+            gubState.lastMountAttempt = esp_timer_get_time();
+            ESP_LOGW(TAG, "SD card not mounted! Attempting to remount...");
+            GUBMountSDCard(SD_CARD_BASE_PATH);
+        }
+
+        canDriverUpdate();
+        canLoggerUpdate();
+        
+        CANMessage_t message;
+        if(canReceiveMessage(&message, pdMS_TO_TICKS(5))){
+            canLoggerProcessMessage(&message);
+        }
+        
+        //* needed if not relying to the timeout of the queue
+        // vTaskDelay(pdMS_TO_TICKS(5));
+    }
+}
+
+/**
+ * Setup the heartbeat led
+*/
+void GUBInitLED(){
+    gubState.lastHeartbeat = 0;
+    #if LED_IS_NEOPIXEL==1
+    ESP_LOGI(TAG, "Heartbeat with RGB");
+    led_strip_config_t strip_config = {
+        .strip_gpio_num = PIN_NUM_HEARTBEAT,
+        .max_leds = 1, // at least one LED on board
+    };
+    led_strip_rmt_config_t rmt_config = {
+        .resolution_hz = 10 * 1000 * 1000, // 10MHz
+    };
+    ESP_ERROR_CHECK(led_strip_new_rmt_device(&strip_config, &rmt_config, &gubState.ledStrip));
+    /* Set all LED off to clear all pixels */
+    led_strip_clear(gubState.ledStrip);
+
+    #else
+    ESP_LOGI(TAG, "Heartbeat with GPIO");
+    gpio_reset_pin(PIN_NUM_HEARTBEAT);
+    /* Set the GPIO as a push/pull output */
+    gpio_set_direction(PIN_NUM_HEARTBEAT, GPIO_MODE_OUTPUT);
+    #endif
+}
+
+/**
+ * Toggle the state of the heartbeat LED
+*/
+void GUBToggleLED(){
+    #if LED_IS_NEOPIXEL==1
+    /* If the addressable LED is enabled */
+    if (gubState.ledState) {
+        /* Set the LED pixel using RGB from 0 (0%) to 255 (100%) for each color */
+        led_strip_set_pixel(gubState.ledStrip, 0, 0, 16, 0);
+        /* Refresh the strip to send data */
+        led_strip_refresh(gubState.ledStrip);
+    } else {
+        /* Set all LED off to clear all pixels */
+        led_strip_clear(gubState.ledStrip);
+    }
+    #else
+
+    gpio_set_level(PIN_NUM_HEARTBEAT, gubState.ledState);
+    #endif
+    
+    gubState.ledState = !gubState.ledState;
+}
+
+/**
+ * Update heartbeat led
+*/
+void GUBHeartbeatUpdate(){
```

Comment:
canLoggerUpdate could probably be spun off to another task, but this was intentionally done so if the main task hung you would physically ba able to see it. This is also the reason for the faster heartbeat period in the config so that small variations are noticable.

Reply:
Can you explain that in a comment somewhere?